File: 位运算/offer-56-1.py
from typing import List


class Solution:
    def singleNumbers(self, nums: List[int]) -> List[int]:
        x, y, m, n = 0, 0, 1, 0
        for num in nums:
            n = n ^ num
        while(n & m == 0):
            m = m << 1
        for num in nums:
            if num & m == 0:
                x = x ^ num
            elif num & m != 0:
                y = y ^ num
        return [x, y]


nums = [1, 2, 10, 4, 1, 4, 3, 3]
obj = Solution()
print(obj.singleNumbers(nums))

File: 位运算/offer-56-1.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    vector<int> singleNumbers(vector<int> &nums)
    {
        int x = 0, y = 0, m = 1, n = 0;
        for (auto num : nums)
            n = n ^ num;
        while ((n & m) == 0) //C++中==优先度高于逻辑运算
            m = m << 1;
        for (auto num : nums){
            if((num&m)==0)
                x = x^num;
            else
                y = y^num;
        return {x,y}; //不使用vector可以减少时间和空间
        }
    }
};